backend/ethereum/channel/contractbackend_test.go

```diff
@@ -100,3 +103,109 @@ func Test_NewWatchOpts(t *testing.T) {
+		Gas:      test.GasLimit,
+		GasPrice: big.NewInt(test.GasPrice),
+		Data:     nil,
+	})
+	opts, err := s.CB.NewTransactor(ctx, 1, s.TxSender.Account)
+	require.NoError(t, err)
+	signed, err := opts.Signer(s.TxSender.Account.Address, rawTx)
+	require.NoError(t, err)
+
+	// Send the TX.
+	require.NoError(t, s.SimBackend.SimulatedBackend.SendTransaction(ctx, signed))
+
+	// Write receipt into `confirmed` when the TX is confirmed.
+	confirmed := make(chan *types.Receipt)
+	go func() {
+		// Confirm.
+		r, err := s.CB.ConfirmTransaction(ctx, signed, s.TxSender.Account)
+		require.NoError(t, err)
+		confirmed <- r
+	}()
+
+	// Create new blocks.
+	for i := 0; i < int(ethchannel.TxFinalityDepth); i++ {
+		// Check that it is not yet confirmed.
+		select {
+		case <-time.After(100 * time.Millisecond):
+		case <-confirmed:
+			t.Error("TX should not be confirmed yet.")
+			t.FailNow()
+		}
+		// Mine new block.
+		s.SimBackend.Commit()
+	}
+	// Wait for confirm.
+	r := <-confirmed
+	// Get current block height.
+	h, err := s.CB.BlockByNumber(ctx, nil)
+	require.NoError(t, err)
+	// Assert that it got included in `TxBlockFinality` many blocks.
+	assert.Equal(t, ethchannel.TxFinalityDepth, (h.NumberU64()-r.BlockNumber.Uint64())+1)
+}
+
+// Test_ReorgConfirmTransaction tests that a TX if confirmed correctly after a
+// reorg.
+func Test_ReorgConfirmTransaction(t *testing.T) {
+	// Test does not make sense for Finality < 2.
+	require.Greater(t, ethchannel.TxFinalityDepth, uint64(1))
+	rng := pkgtest.Prng(t)
+	s := test.NewSimSetup(rng)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Create the Transaction.
+	rawTx := types.NewTx(&types.LegacyTx{
+		Nonce:    0,
+		To:       &common.Address{},
+		Value:    big.NewInt(1),
+		Gas:      test.GasLimit,
+		GasPrice: big.NewInt(test.GasPrice),
+		Data:     nil,
+	})
+	opts, err := s.CB.NewTransactor(ctx, 1, s.TxSender.Account)
+	require.NoError(t, err)
+	tx, err := opts.Signer(s.TxSender.Account.Address, rawTx)
+	require.NoError(t, err)
+
+	// Send the TX.
+	require.NoError(t, s.SimBackend.SimulatedBackend.SendTransaction(ctx, tx))
+	// Wait `TxFinalityDepth - 1` many blocks.
+	for i := uint64(0); i < ethchannel.TxFinalityDepth-1; i++ {
+		s.SimBackend.Commit()
+	}
+
+	// Check that the TX is not confirmed yet.
+	short, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	_, err = s.CB.ConfirmTransaction(short, tx, s.TxSender.Account)
+	defer cancel()
+	require.Error(t, err)
+
+	// Do a reorg and add two more blocks. Move the TX one block forward.
+	// The TX should now be included in `TxFinalityDepth` many blocks.
+	s.SimBackend.Reorg(ctx, ethchannel.TxFinalityDepth-1, func(txs []types.Transactions) []types.Transactions {
+		ret := make([]types.Transactions, ethchannel.TxFinalityDepth+1)
+		ret[1] = txs[0]
+		return ret
+	})
+
+	// Confirm
+	_, err = s.CB.ConfirmTransaction(ctx, tx, s.TxSender.Account)
+	require.NoError(t, err)
+}
```

Comment:
Can you add a test asserting that a transaction is not confirmed if it is dropped by a reorg before the finality depth is reached?

Reply:
Ok i will try.

Reply:
I added a test and simplified `Test_ReorgConfirmTransaction` to use the `TokenSetup` which was not invented as the test was originally written.